homeassistant/components/tuya.py

```diff
@@ -0,0 +1,166 @@
+            dev_type = device.device_type()
+            if (dev_type in TUYA_TYPE_TO_HA and
+                    device.object_id() not in hass.data[DOMAIN]['entities']):
+                ha_type = TUYA_TYPE_TO_HA[dev_type]
+                if ha_type not in device_type_list:
+                    device_type_list[ha_type] = []
+                device_type_list[ha_type].append(device.object_id())
+        for ha_type, dev_ids in device_type_list.items():
+            discovery.load_platform(
+                hass, ha_type, DOMAIN,
+                {'dev_ids': dev_ids}, config)
+
+    device_list = tuya.get_all_devices()
+    load_devices(device_list)
+
+    def poll_devices_update(event_time):
+        """Check if accesstoken is expired and pull device list from server."""
+        _LOGGER.info("Pull Devices From Tuya")
+        tuya.poll_devices_update()
+        # Add new discover device.
+        device_list = tuya.get_all_devices()
+        load_devices(device_list)
+        # Delete not exist device.
+        newlist_ids = []
+        for device in device_list:
+            newlist_ids.append(device.object_id())
+        for dev_id in list(hass.data[DOMAIN]['entities']):
+            if dev_id not in newlist_ids:
+                dispatcher_send(hass, SIGNAL_DELETE_ENTITY, dev_id)
+                hass.data[DOMAIN]['entities'].pop(dev_id)
+
+    track_time_interval(hass, poll_devices_update, timedelta(minutes=5))
+
+    hass.services.register(DOMAIN, SERVICE_PULL_DEVICES, poll_devices_update)
+
+    def force_update(call):
+        """Force all devices to pull data."""
+        _LOGGER.info("Refreshing Device Data From Tuya")
+        dispatcher_send(hass, SIGNAL_UPDATE_ENTITY)
+
+    hass.services.register(DOMAIN, SERVICE_FORCE_UPDATE, force_update)
+
+    return True
+
+
+class TuyaDevice(Entity):
+    """Tuya base device."""
+
+    def __init__(self, tuya):
+        """Init Tuya devices."""
+        self.tuya = tuya
+
+    async def async_added_to_hass(self):
+        """Call when entity is added to hass."""
+        dev_id = self.tuya.object_id()
+        self.hass.data[DOMAIN]['entities'][dev_id] = self.entity_id
+        async_dispatcher_connect(
+            self.hass, SIGNAL_DELETE_ENTITY, self._delete_callback)
+        async_dispatcher_connect(
+            self.hass, SIGNAL_UPDATE_ENTITY, self._update_callback)
+
+    @property
+    def object_id(self):
+        """Return Tuya device id."""
+        return self.tuya.object_id()
+
+    @property
+    def name(self):
+        """Return Tuya device name."""
+        return self.tuya.name()
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self.tuya.state()
+
+    @property
+    def icon(self):
+        """Return the entity picture to use in the frontend, if any."""
+        return self.tuya.iconurl()
+
+    @property
+    def available(self):
+        """Return if the device is available."""
+        return self.tuya.available()
```

Comment:
Is this call doing any I/O? I/O is not permitted within entity properties, since they will be evaluated in the event loop. I'm asking in general about all the calls in the entity properties.

Reply:
No, it is a state about the device is online or offline. It refresh while update called.

Reply:
If the source code for tuyapy is necessary? I didn't upload it.